### PR TITLE
[TK-09295] Bootstrap Queries as loop at space level

### PR DESCRIPTION
**Pull request by neonphog**
DEPENDS ON https://github.com/holochain/holochain/pull/817

- previously every agent joined to a space would start a bootstrap loop
- previously requests would randomly also decide to query the bootstrap server
- now we have a single loop at the space level with an exponential backoff

### CODE REVIEW

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs [hunk 1]
```diff
@@ -495,6 +467,59 @@ impl Space {
...
+                                if let Ok(is_local) = i_s_c.is_agent_local(agent.clone()).await {
+                                    if !is_local {
+                                        // we got a result - let's add it to our store for the future
+                                        let _ = evt_s_c
+                                            .put_agent_info_signed(PutAgentInfoSignedEvt {
+                                                space: space_c.clone(),
+                                                agent,
+                                                agent_info_signed: item.clone(),
+                                            })
+                                            .await;
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:513]: Let's log potential errors?

Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs]: done

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs [hunk 2]
```diff
@@ -495,6 +467,65 @@ impl Space {
...
+            let bootstrap_service = config.bootstrap_service.clone();
+            let space_c = space.clone();
+            tokio::task::spawn(async move {
+                const START_DELAY: std::time::Duration = std::time::Duration::from_secs(1);
+                const MAX_DELAY: std::time::Duration = std::time::Duration::from_secs(60 * 60);
+
+                let mut delay_len = START_DELAY;
+
+                loop {
+                    tokio::time::sleep(delay_len).await;
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:483]: So, this loop never ends, and exponentially backs off to one minute, at which point it pings the bootstrap server every minute forever?

I know this isn't part of the change in this PR. Seems fine for now but maybe we need to change this once we expect more Holochain users? If we have 100k conductors, that's 1600 requests per second total to the bootstrap server.

**Review by maackle — APPROVED** [review]
Looks good, just a few remarks on discarded errors, and a more long term question

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs (hunk 2 above)
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:483]: It backs off to 1 hour, then pings it once per hour. I think once per hour is reasonable.

Comment (maackle) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:483]: Oops, yeah. Rushed that last message.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs [hunk 3]
```diff
@@ -254,37 +251,6 @@ pub(crate) fn get_5_or_less_non_local_agents_near_basis(
             }
```
[the remark this answers is not in the document]
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs]: :+1: done

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs [hunk 4]
```diff
@@ -495,6 +467,65 @@ impl Space {
...
+                    .await
+                    {
+                        Err(e) => {
+                            tracing::error!(msg = "Failed to get peers from bootstrap", ?e);
+                        }
+                        Ok(list) => {
+                            for item in list {
+                                // TODO - someday some validation here
+                                let agent = item.agent.clone();
+                                if let Ok(is_local) = i_s_c.is_agent_local(agent.clone()).await {
```
[the remark this answers is not in the document]
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs]: :+1: done

**Pull request merged**